File: app/src/main/java/com/icostel/arhitecturesample/utils/AppExecutors.java
package com.icostel.arhitecturesample.utils;

import android.os.Handler;
import android.os.Looper;

import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

import javax.inject.Inject;
import javax.inject.Singleton;

import androidx.annotation.NonNull;

@Singleton
public class AppExecutors {

    private static final int NETWORK_THREADS = 3;

    private Executor diskIO;
    private Executor networkIO;
    private Executor mainThread;
    private Executor background;

    @Inject
    AppExecutors() {
        diskIO = Executors.newSingleThreadExecutor();
        networkIO = Executors.newFixedThreadPool(NETWORK_THREADS);
        mainThread = new MainThreadExecutor();
        background = Executors.newSingleThreadExecutor();
    }

    public Executor diskIO() {
        return diskIO;
    }

    public Executor networkIO() {
        return networkIO;
    }

    public Executor mainThread() {
        return mainThread;
    }

    public Executor background() {
        return background;
    }

    private static class MainThreadExecutor implements Executor {
        private Handler mainThreadHandler = new Handler(Looper.getMainLooper());

        @Override
        public void execute(@NonNull Runnable runnable) {
            mainThreadHandler.post(runnable);
        }
    }
}

File: commons/src/main/java/com/icostel/commons/utils/AppExecutors.kt
package com.icostel.commons.utils

import android.os.Handler
import android.os.Looper

import java.util.concurrent.Executor
import java.util.concurrent.Executors

import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class AppExecutors @Inject
internal constructor() {

    private val diskIO: Executor
    private val networkIO: Executor
    private val mainThread: Executor
    private val background: Executor

    init {
        diskIO = Executors.newSingleThreadExecutor()
        networkIO = Executors.newFixedThreadPool(NETWORK_THREADS)
        background = Executors.newSingleThreadExecutor()
        mainThread = MainThreadExecutor()
    }

    fun diskIO(): Executor {
        return diskIO
    }

    fun networkIO(): Executor {
        return networkIO
    }

    fun mainThread(): Executor {
        return mainThread
    }

    fun background(): Executor {
        return background
    }

    private class MainThreadExecutor : Executor {
        private val mainThreadHandler = Handler(Looper.getMainLooper())

        override fun execute(runnable: Runnable) {
            mainThreadHandler.post(runnable)
        }
    }

    companion object {
        private const val NETWORK_THREADS = 3
    }
}

